cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc

```diff
@@ -391,6 +391,49 @@ struct CastFunctor<O, I,
   }
 };
 
+// ----------------------------------------------------------------------
+// Integer to decimal
+
+struct IntegerToDecimal {
+  template <typename OutValue, typename IntegerType>
+  OutValue Call(KernelContext*, IntegerType val, Status* st) const {
+    auto maybe_decimal = OutValue(val).Rescale(0, out_scale_);
+    if (ARROW_PREDICT_TRUE(maybe_decimal.ok())) {
+      return maybe_decimal.MoveValueUnsafe();
+    }
+    *st = maybe_decimal.status();
+    return OutValue{};
+  }
+
+  int32_t out_scale_;
+};
+
+template <typename O, typename I>
+struct CastFunctor<O, I,
+                   enable_if_t<is_decimal_type<O>::value && is_integer_type<I>::value>> {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const auto& out_type = checked_cast<const O&>(*out->type());
+    const auto out_scale = out_type.scale();
+    const auto out_precision = out_type.precision();
+
+    // verify precision and scale
+    if (out_scale < 0) {
+      return Status::Invalid("Scale must be non-negative");
+    }
+    // maximal number of decimal digits for int8/16/32/64
+    constexpr std::array<int, 4> decimal_digits{3, 5, 10, 19};
```

Comment:
Since `decimal_digits` is a constant array, I would make it a `static` data member of `CastFunctor`. Also, you can use a standard array where the size would be implicit and is more lightweight than `std::array`:
```c++
static constexpr int decimal_digits[]{3,5, 10, 19};
```

Reply:
`static` will put the array in global data section, without `static` the array is in code section nearby. Look `static` is not good here.
But I guess we don't need to worry about it. For this trivial code snippet, compiler should be able to figure out the index and throw away this table.

Reply:
Agree, local variable is better in this scenario.